source/extensions/transport_sockets/tls/BUILD

```diff
@@ -10,17 +11,52 @@ licenses(["notice"])  # Apache 2
 
 envoy_extension_package()
 
-envoy_cc_extension(
-    name = "config",
-    srcs = ["config.cc"],
+envoy_cc_library(
+    name = "base_config",
     hdrs = ["config.h"],
-    # TLS is core functionality.
-    visibility = ["//visibility:public"],
     deps = [
         "//envoy/network:transport_socket_interface",
         "//envoy/registry",
         "//envoy/server:transport_socket_config_interface",
         "//source/common/tls:ssl_socket_lib",
+    ],
+)
+
+envoy_cc_library(
+    name = "downstream_config",
+    srcs = ["downstream_config.cc"],
+    hdrs = ["downstream_config.h"],
+    deps = [
+        ":base_config",
+        "//envoy/network:transport_socket_interface",
+        "//envoy/registry",
+        "//envoy/server:transport_socket_config_interface",
+        "//source/common/tls:ssl_socket_lib",
         "@envoy_api//envoy/extensions/transport_sockets/tls/v3:pkg_cc_proto",
     ],
+    alwayslink = True,
+)
+
+envoy_cc_library(
+    name = "upstream_config",
+    srcs = ["upstream_config.cc"],
+    hdrs = ["upstream_config.h"],
+    deps = [
+        ":base_config",
+        "//envoy/network:transport_socket_interface",
+        "//envoy/registry",
+        "//envoy/server:transport_socket_config_interface",
+        "//source/common/tls:ssl_socket_lib",
+        "@envoy_api//envoy/extensions/transport_sockets/tls/v3:pkg_cc_proto",
+    ],
+    alwayslink = True,
+)
+
+envoy_cc_extension(
```

Comment:
I'm intentionally leaving a config library which includes both because for legacy servers we always had them under one extension, and there's a lot of build target churn otherwise.  I can do the churn if you think it's worth it